lib/_http_common.js

```diff
@@ -77,6 +77,20 @@ function parserOnHeadersComplete(versionMajor, versionMinor, headers, method,
     parser.incoming.statusMessage = statusMessage;
   }
 
+  // The client made non-upgrade request, and server is just advertising
+  // supported protocols.
+  //
+  // See RFC7230 Section 6.7
+  //
+  // NOTE: RegExp below matches `upgrade` in `Connection: abc, upgrade, def`
+  // header.
+  if (upgrade &&
```

Comment:
Isn't it http-parser which sets the `upgrade` argument based on the response header [her](https://github.com/nodejs/http-parser/blob/4e382f96e6d3321538a78f2c7f9506d4e79b08d6/http_parser.c#L1817-L1820)?

And yeah, you require both header on the request. I still think the upgrade decision could be solely based on the response headers.